src/common/wrap/wrap-websocket.js

```diff
@@ -0,0 +1,72 @@
+import { globalScope } from '../constants/runtime'
+import { originals } from '../config/state/originals'
+import { now } from '../timing/now'
+import { checkState } from '../window/load'
+import { generateRandomHexString } from '../ids/unique-id'
+
+export const WEBSOCKET_TAG = 'websocket-'
+export const ADD_EVENT_LISTENER_TAG = 'addEventListener'
+
+const wrapped = {}
+
+export function wrapWebSocket (sharedEE) {
+  if (wrapped[sharedEE.debugId]++) return sharedEE
+  if (!originals.WS) return sharedEE
+
+  function reporter (socketId) {
+    const createdAt = now()
+    return function (message, ...data) {
+      const timestamp = data[0]?.timeStamp || now()
+      const isLoaded = checkState()
+      sharedEE.emit(WEBSOCKET_TAG + message, [timestamp, timestamp - createdAt, isLoaded, socketId, ...data])
+    }
+  }
+
+  Object.defineProperty(WrappedWebSocket, 'name', {
+    value: 'WebSocket'
+  })
+
+  function WrappedWebSocket () {
```

Comment:
I wonder if we could use class inheritance here. For example `window.WebSocket = class extends WebSocket`